frame/contracts/src/lib.rs

```diff
@@ -949,10 +897,18 @@ pub mod pallet {
 	/// Evicted contracts that await child trie deletion.
 	///
 	/// Child trie deletion is a heavy operation depending on the amount of storage items
-	/// stored in said trie. Therefore this operation is performed lazily in `on_initialize`.
+	/// stored in said trie. Therefore this operation is performed lazily in `on_idle`.
+	#[pallet::storage]
+	pub(crate) type DeletionQueueMap<T: Config> = StorageMap<_, Twox64Concat, u32, TrieId>;
+
+	/// A pair of monotonic counters used to track the latest contract marked for deletion
+	/// and the latest deleted contract in DeletionQueueMap.
+	///
+	/// When we iterate the map to remove contracts, we simply use the `delete_nonce` counter and
+	/// don't pay the cost of an extra call to `sp_io::storage::next_key` to lookup the next entry
+	/// in the map
 	#[pallet::storage]
-	pub(crate) type DeletionQueue<T: Config> =
-		StorageValue<_, BoundedVec<DeletedContract, T::DeletionQueueDepth>, ValueQuery>;
+	pub(crate) type DeletionQueueNonces<T: Config> = StorageValue<_, DeletionQueue<T>, ValueQuery>;
```

Comment:
It is not really a nonce because of the wrap around.

```suggestion
	pub(crate) type DeletionQueueCounter<T: Config> = StorageValue<_, DeletionQueue<T>, ValueQuery>;
```